File: src/lib/protocol.ts
/** Mirror of `core/src/protocol.rs`. Keep the two files in sync. */
export interface AppInfo {
  name: string;
  version: string;
}

/**
 * Where the backend runs. `local` is the engine in the app's own process;
 * the others are a daemon reached over `wsl.exe` or `ssh`, because watching
 * files across a network filesystem does not work.
 */
export type ConnectionTarget =
  { kind: "local" } | { kind: "wsl"; distro: string } | { kind: "ssh"; host: string };

/** Liveness of the backend connection. */
export type ConnectionState =
  | "connecting"
  /** Putting a daemon on a remote that hasn't got one. */
  | "installing"
  | "connected"
  | "disconnected"
  | "failed";

/** The current backend connection, surfaced in the status bar. */
export interface ConnectionInfo {
  target: ConnectionTarget;
  state: ConnectionState;
  label: string;
  /** True when the files are not on this machine. */
  remote: boolean;
  /** Why, when the state is `disconnected` or `failed`. */
  message: string | null;
  /** From the handshake, once it has happened. */
  daemonVersion: string | null;
  /**
   * Feature names the daemon reported. Empty means unknown — assume nothing
   * (local, or an older daemon that never sent the field).
   */
  capabilities: string[];
  /**
   * True when a remote daemon's package version differs from this app's.
   * The connection still works; the UI surfaces the fact. Always false local.
   */
  daemonStale: boolean;
  /** Unix epoch ms of the last state change; the feed's gap marker uses it. */
  since: number;
}

/** The currently open workspace. */
export interface WorkspaceInfo {
  /** Absolute path, forward slashes, for display. */
  root: string;
  /** Final path component. */
  name: string;
  /** Unix epoch milliseconds. */
  watchingSince: number;
}

/** One entry in a single directory listing. */
export interface DirEntryNode {
  name: string;
  /** Workspace-relative, forward slashes. */
  path: string;
  isDir: boolean;
  /**
   * True when git ignores this entry. An ignored entry only appears at all
   * with `showIgnored` on, or when a pin / agent context file forces it.
   */
  ignored: boolean;
  /** A file that instructs a coding agent, surfaced as one. */
  agentContext: boolean;
}

/** One entry of the workspace's pinned list, resolved against the disk. */
export interface PinnedEntry {
  /** Workspace-relative, forward slashes — exactly as stored in settings. */
  path: string;
  /** Final path component, for display. */
  name: string;
  isDir: boolean;
  /** False for a pin whose target has been renamed or deleted. */
  exists: boolean;
}

/**
 * One directory offered by the folder picker. Absolute, unlike everything
 * else here, because it is chosen *before* there is a workspace to be
 * relative to.
 */
export interface BrowseEntry {
  name: string;
  /** Absolute path on the backend's machine, forward slashes. */
  path: string;
  /** This directory is a git repository — the most useful thing to know when
   * picking one out of a list of twenty. */
  isRepository: boolean;
}

/** One directory's worth of the folder picker. */
export interface BrowseListing {
  /** Absolute path of the directory that was listed. */
  path: string;
  /** `null` at the filesystem root, which is where "go up" stops. */
  parent: string | null;
  entries: BrowseEntry[];
  /** The directory held more than the listing cap. */
  truncated: boolean;
}

/** How git sees one file. */
export type GitStatusKind =
  "added" | "modified" | "deleted" | "renamed" | "untracked" | "conflicted";

/** Git status of a single file. */
export interface GitFileStatus {
  /** Workspace-relative, forward slashes. */
  path: string;
  status: GitStatusKind;
  /** True if the change is in the index. */
  staged: boolean;
}

/** A full git status snapshot for the open workspace. */
export interface GitStatusSnapshot {
  isRepository: boolean;
  /** `null` when detached or unborn. */
  branch: string | null;
  files: GitFileStatus[];
}

/**
 * Whether git mutations can be offered for the open workspace. Checked on
 * open so the UI degrades to read-only with a hint rather than showing
 * buttons that fail when pressed.
 */
export interface GitCapabilities {
  canMutate: boolean;
  /** `git --version` output, when it ran. */
  version: string | null;
  /** Why mutations are unavailable, for the hint. */
  reason: string | null;
}

/** Local branches and which one is checked out. */
export interface BranchList {
  /** `null` on a detached HEAD or an unborn branch. */
  current: string | null;
  branches: string[];
}

/** Kind of filesystem change reported by the watcher. */
export type FsEventKind = "created" | "modified" | "deleted" | "renamed";

/** One filesystem change, already debounced, filtered, and normalized. */
export interface FsEvent {
  kind: FsEventKind;
  /** Workspace-relative, forward slashes. */
  path: string;
  /** Best-effort; false if the path no longer exists. */
  isDir: boolean;
  /** Unix epoch milliseconds. */
  at: number;
}

/**
 * An `FsEvent` joined to the agent tool call that most likely caused it.
 * `attribution` is `null` when nothing claimed the change — the user's own
 * editor, or a write the correlation window could not safely tie to a call.
 */
export interface AttributedEvent {
  event: FsEvent;
  /** `null` when nothing claimed it — an external edit. */
  attribution: Attribution | null;
}

/** Which agent tool call is believed to have caused a filesystem change. */
export interface Attribution {
  sessionId: string;
  agent: AgentKind;
  tool: string;
  /** The agent's one-line intent, when the provider supplied one. */
  summary: string | null;
  /**
   * True when claimed by a shell command rather than an explicit path —
   * lower confidence, and the UI says so.
   */
  viaCommand: boolean;
  /** Delegated subagent work. Absent on events from an older daemon. */
  sidechain?: boolean;
}

/** Lifecycle state of the filesystem watcher for the open workspace. */
export type WatcherState = "off" | "running" | "error";

/** Current watcher state, surfaced in the status bar. */
export interface WatcherStatus {
  state: WatcherState;
  /** Populated on `error`. */
  message: string | null;
}

/**
 * What the preview pane should render for one file. Discriminated on `kind`,
 * mirroring the Rust tagged enum.
 */
export type PreviewPayload =
  | { kind: "text"; path: string; text: string; language: string }
  | { kind: "image"; path: string; mime: string; base64: string }
  | { kind: "binary"; path: string; size: number }
  | { kind: "tooLarge"; path: string; size: number };

/** Why a file comparison (session or git) can't be produced. */
export type DiffUnavailable = "notARepository" | "notTracked" | "notText" | "tooLarge";

/** The two sides of a file comparison — session or git. */
export interface SessionDiff {
  path: string;
  /** Left-hand content; `null` if the file didn't exist on that side. */
  baseline: string | null;
  /** Right-hand content; `null` if the file is absent on that side. */
  current: string | null;
  /** Set when no meaningful comparison exists; null means the diff is usable. */
  unavailable: DiffUnavailable | null;
}

/** Settings that apply to one workspace, persisted per root. */
export interface WorkspaceSettings {
  /** Extra gitignore-syntax globs, hidden from tree, index, and feed. */
  extraIgnores: string[];
  /** Show git-ignored entries in the tree and file index — never the feed. */
  showIgnored: boolean;
  /** Paths kept visible whatever `.gitignore` says, grouped at the tree top. */
  pinned: string[];
}

/** Settings that apply to every workspace, persisted once for the app. */
export interface AppSettings {
  /** Surface `AGENTS.md` and friends whatever `.gitignore` says. */
  showAgentContext: boolean;
  /**
   * Extra directories to search for agent sessions on the connected machine,
   * added to whatever the app detects itself. Persisted per host so a local
   * path is not sent to a remote daemon.
   */
  agentRoots: string[];
  /**
   * What to run on the far side of a WSL or SSH connection. A bare name works
   * when the daemon is on the remote `PATH`; an SSH command runs without a
   * login shell, so an absolute path is often needed.
   */
  daemonCommand: string;
  /**
   * Install the daemon on a remote that hasn't got one, rather than failing
   * with instructions. On by default.
   */
  autoInstallDaemon: boolean;
  /**
   * Max activity-feed batches kept in the UI (oldest drop off). Default 250.
   * Presentation only — the backend ignores this.
   */
  feedMaxEntries: number;
  /**
   * Check GitHub for a newer release at startup. Notify-only — nothing is
   * ever downloaded or installed.
   */
  checkForUpdates: boolean;
  /**
   * OS notification when an agent waits or finishes. Only while unfocused.
   * Presentation only — the backend ignores this.
   */
  notifyAgentState: boolean;
}

/** Result of a notify-only release check. */
export interface UpdateCheck {
  /** The running version. */
  current: string;
  /**
   * Newest published release tag, without the leading `v`; `null` when the
   * check was skipped, failed, or found nothing.
   */
  latest: string | null;
  /** Where to read about it. */
  url: string | null;
  /** True only when `latest` is strictly newer than `current`. */
  newer: boolean;
}

/** Which coding agent a session belongs to. */
export type AgentKind = "claudeCode" | "grok";

/**
 * What an agent is doing right now, normalized across providers.
 * Discriminated on `kind`, mirroring the Rust tagged enum.
 *
 * `detail` on `working` is display-only (e.g. "thinking", a tool name).
 */
export type AgentActivity =
  | { kind: "working"; detail: string | null }
  | { kind: "blocked" }
  | { kind: "idle" }
  | { kind: "stale" };

/** One directory the app searches for agent sessions, as shown in settings. */
export interface AgentRootInfo {
  /** Absolute path, forward slashes — these live outside any workspace. */
  path: string;
  /**
   * Which agent recognises this directory. `null` means none did: for a path
   * the user typed, that is why they see no sessions, so it must be shown.
   */
  agent: AgentKind | null;
  /** Found automatically, as opposed to named by the user. */
  detected: boolean;
  /**
   * Why detection found nothing, when this row is a diagnostic rather than
   * a folder. Empty `path` plus a `note` is the empty-list case.
   */
  note?: string | null;
}

/** A session the app has found on disk but isn't necessarily tailing yet. */
export interface SessionRef {
  /** Provider-assigned id, unique per provider. */
  id: string;
  agent: AgentKind;
  /** Generated session title, when the provider supplies one. */
  title: string | null;
  /** Unix epoch milliseconds of the most recent record seen. */
  lastActivity: number;
  /** What the agent is doing right now, as best the provider can say. */
  activity: AgentActivity;
}

/**
 * One thing an agent did, normalized across providers. Discriminated on
 * `kind`, mirroring the Rust tagged enum.
 */
export type AgentEvent =
  | {
      kind: "sessionStarted";
      sessionId: string;
      agent: AgentKind;
      title: string | null;
      at: number;
    }
  | {
      kind: "toolCall";
      sessionId: string;
      /** Absent on events from an older daemon of this protocol version. */
      agent?: AgentKind;
      at: number;
      tool: string;
      /** One-line description of what the call was for, when derivable. */
      summary: string | null;
      /** Workspace-relative, forward slashes. */
      paths: string[];
      /** The work of a subagent rather than the main thread. */
      sidechain: boolean;
    }
  | {
      kind: "assistantNote";
      sessionId: string;
      agent?: AgentKind;
      at: number;
      text: string;
    }
  | {
      kind: "activityChanged";
      sessionId: string;
      agent?: AgentKind;
      at: number;
      activity: AgentActivity;
    }
  | { kind: "sessionEnded"; sessionId: string; agent?: AgentKind; at: number };

/** The result of tailing a session: what happened, plus what couldn't be read. */
export interface AgentPoll {
  events: AgentEvent[];
  /** Records read since the workspace opened. */
  records: number;
  /** Of those, how many could not be parsed. */
  skipped: number;
}

/**
 * Event names emitted by the backend. Rust `emit` calls and these `listen`
 * calls must both go through the shared constants so the names can't
 * silently drift apart (mirrors `src-tauri/src/protocol.rs`).
 */
export const EVENT_FS_CHANGES = "fs-changes";
export const EVENT_GIT_STATUS = "git-status";
export const EVENT_WATCHER_STATUS = "watcher-status";
/** Filesystem changes with optional agent attribution. */
export const EVENT_ATTRIBUTED = "attributed-changes";
/** Normalized agent activity from the background poller. */
export const EVENT_AGENT_EVENTS = "agent-events";
export const EVENT_CONNECTION = "connection";

File: src-tauri/src/protocol.rs
//! Serializable types crossing the UI <-> backend boundary.
//!
//! Everything the frontend sends or receives is defined here and mirrored in
//! `src/lib/protocol.ts`. This is deliberate: a later phase replaces the
//! in-process backend with a remote daemon, and only serializable messages
//! survive that move.

use serde::{Deserialize, Serialize};

/// Identity of the running application, surfaced in the window title and UI.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AppInfo {
    pub name: String,
    pub version: String,
}

/// The result type for every `#[tauri::command]`, so the error shape is
/// uniform across the UI <-> backend boundary.
pub type CommandResult<T> = Result<T, String>;

/// The currently open workspace.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkspaceInfo {
    /// Absolute path, forward slashes, for display.
    pub root: String,
    /// Final path component.
    pub name: String,
    /// Unix epoch milliseconds.
    pub watching_since: i64,
}

/// One entry in a single directory listing.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DirEntryNode {
    pub name: String,
    /// Workspace-relative, forward slashes.
    pub path: String,
    pub is_dir: bool,
    /// True when git ignores this entry. An ignored entry only appears at all
    /// when `show_ignored` is on or something forces it visible (a pin, or an
    /// agent context file).
    pub ignored: bool,
    /// This is a file that instructs a coding agent, and the app is surfacing
    /// it as one. False when the app-level setting is off.
    pub agent_context: bool,
}

/// One entry of the workspace's pinned list, resolved against the disk.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PinnedEntry {
    /// Workspace-relative, forward slashes — exactly as stored in settings.
    pub path: String,
    /// Final path component, for display.
    pub name: String,
    pub is_dir: bool,
    /// False for a pin whose target has been renamed or deleted. Dead pins are
    /// shown rather than dropped: a pin vanishing on its own is worse than a
    /// visible stale one.
    pub exists: bool,
}

/// How git sees one file.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum GitStatusKind {
    Added,
    Modified,
    Deleted,
    Renamed,
    Untracked,
    Conflicted,
}

/// Git status of a single file.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitFileStatus {
    /// Workspace-relative, forward slashes.
    pub path: String,
    pub status: GitStatusKind,
    /// True if the change is in the index.
    pub staged: bool,
}

/// A full git status snapshot for the open workspace.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitStatusSnapshot {
    pub is_repository: bool,
    /// `None` when detached or unborn.
    pub branch: Option<String>,
    pub files: Vec<GitFileStatus>,
}

/// Kind of filesystem change reported by the watcher.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum FsEventKind {
    Created,
    Modified,
    Deleted,
    Renamed,
}

/// One filesystem change, already debounced, filtered, and normalized.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FsEvent {
    pub kind: FsEventKind,
    /// Workspace-relative, forward slashes.
    pub path: String,
    /// Best-effort; false if the path no longer exists.
    pub is_dir: bool,
    /// Unix epoch milliseconds.
    pub at: i64,
}

/// Lifecycle state of the filesystem watcher for the open workspace.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum WatcherState {
    Off,
    Running,
    Error,
}

/// Current watcher state, surfaced in the status bar.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WatcherStatus {
    pub state: WatcherState,
    /// Populated on `Error`.
    pub message: Option<String>,
}

/// What the preview pane should render for one file. Tagged by `kind` so the
/// TS mirror is a discriminated union.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "camelCase")]
pub enum PreviewPayload {
    /// Decoded UTF-8 text, with a Shiki language id inferred from the
    /// extension (`"text"` when nothing matches).
    Text {
        path: String,
        text: String,
        language: String,
    },
    /// Base64-encoded image bytes; the UI builds a `data:` URL.
    Image {
        path: String,
        mime: String,
        base64: String,
    },
    /// Not decodable as text, so there is nothing read-only to show.
    Binary { path: String, size: u64 },
    /// Over the preview size cap.
    TooLarge { path: String, size: u64 },
}

/// Why a "diff since session" can't be produced.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum DiffUnavailable {
    /// The workspace isn't a git repository, and the baseline comes from git.
    NotARepository,
    /// Git ignores this file, so it has no `HEAD` blob and never appears in
    /// the status output the session baseline is captured from. Showing it as
    /// wholly added would misrepresent a file that may well have existed.
    NotTracked,
}

/// The two sides of a "diff since session" comparison. The line diff itself
/// is computed in the UI, which already has a diff library.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionDiff {
    pub path: String,
    /// Content when the session started; `None` if the file didn't exist.
    pub baseline: Option<String>,
    /// Content now; `None` if the file has since been deleted.
    pub current: Option<String>,
    /// Set when no meaningful comparison exists; `None` means the diff is
    /// usable.
    pub unavailable: Option<DiffUnavailable>,
}

/// Settings that apply to one workspace, persisted per root.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
// `default` so settings persisted by an older version still deserialize
// rather than resetting the whole workspace entry.
#[serde(rename_all = "camelCase", default)]
pub struct WorkspaceSettings {
    /// Extra gitignore-syntax globs, hidden from the tree, the file index,
    /// and the activity feed.
    pub extra_ignores: Vec<String>,
    /// Show entries git ignores in the tree and file index. Deliberately does
    /// not affect the activity feed: hiding ignored churn there is what keeps
    /// an `npm install` from flooding it.
    pub show_ignored: bool,
    /// Workspace-relative paths (files and directories) kept visible whatever
    /// `.gitignore` says, and grouped at the top of the tree.
    pub pinned: Vec<String>,
}

/// Settings that apply to every workspace, persisted once for the app.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
// `default` for the same forward-compatibility reason as `WorkspaceSettings`,
// and because the defaults here are not all `false`.
#[serde(rename_all = "camelCase", default)]
pub struct AppSettings {
    /// Surface the files that instruct a coding agent (`AGENTS.md` and
    /// friends) whatever `.gitignore` says. App-level because "always show me
    /// `AGENTS.md`" is how you work, not a property of one repo.
    pub show_agent_context: bool,
}

impl Default for AppSettings {
    fn default() -> Self {
        AppSettings {
            show_agent_context: true,
        }
    }
}

/// Event names emitted by the backend. Rust `emit` calls and TS `listen`
/// calls must both go through these constants (mirrored in
/// `src/lib/protocol.ts`) so the names can't silently drift apart.
pub const EVENT_FS_CHANGES: &str = "fs-changes";
pub const EVENT_GIT_STATUS: &str = "git-status";
pub const EVENT_WATCHER_STATUS: &str = "watcher-status";

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn app_info_serializes_with_camel_case_fields() {
        let info = AppInfo {
            name: "AgentLens".into(),
            version: "0.0.1".into(),
        };

        let json = serde_json::to_value(&info).unwrap();
        assert_eq!(json["name"], "AgentLens");
        assert_eq!(json["version"], "0.0.1");
    }

    #[test]
    fn fs_event_kind_round_trips_through_camel_case_json() {
        let cases = [
            (FsEventKind::Created, "\"created\""),
            (FsEventKind::Modified, "\"modified\""),
            (FsEventKind::Deleted, "\"deleted\""),
            (FsEventKind::Renamed, "\"renamed\""),
        ];
        for (kind, expected) in cases {
            let json = serde_json::to_string(&kind).unwrap();
            assert_eq!(json, expected);
            let round_tripped: FsEventKind = serde_json::from_str(&json).unwrap();
            assert_eq!(round_tripped, kind);
        }
    }

    #[test]
    fn watcher_state_round_trips_through_camel_case_json() {
        let cases = [
            (WatcherState::Off, "\"off\""),
            (WatcherState::Running, "\"running\""),
            (WatcherState::Error, "\"error\""),
        ];
        for (state, expected) in cases {
            let json = serde_json::to_string(&state).unwrap();
            assert_eq!(json, expected);
            let round_tripped: WatcherState = serde_json::from_str(&json).unwrap();
            assert_eq!(round_tripped, state);
        }
    }

    #[test]
    fn fs_event_serializes_with_camel_case_fields() {
        let event = FsEvent {
            kind: FsEventKind::Modified,
            path: "src/main.rs".into(),
            is_dir: false,
            at: 1_700_000_000_000,
        };
        let json = serde_json::to_value(&event).unwrap();
        assert_eq!(json["kind"], "modified");
        assert_eq!(json["path"], "src/main.rs");
        assert_eq!(json["isDir"], false);
        assert_eq!(json["at"], 1_700_000_000_000i64);
    }

    #[test]
    fn watcher_status_serializes_with_camel_case_fields() {
        let status = WatcherStatus {
            state: WatcherState::Error,
            message: Some("boom".into()),
        };
        let json = serde_json::to_value(&status).unwrap();
        assert_eq!(json["state"], "error");
        assert_eq!(json["message"], "boom");
    }
}
